armi/utils/asciimaps.py

```diff
@@ -40,236 +60,428 @@ class AsciiMap(object):
+        Base: (0,0), (1,0)  (0,1),  (1,1),  (0,2), (-1,3), (0,3), (-1,4), (-2,5), (-1,5), (-2,6), (-3,7) (-2,7)
 
-    Then stride inside your triplet by position: row%3
-    base = (tI-pos, tJ+pos)
+        Looking graphically, there are basically 3 rays going up at 120 degrees.
+        So we can find a consistent pattern for each ray and use a modulus to figure
+        out which ray we're on.
+
+        """
+        if asciiLineNum == 0:
+            return 0, 0
+        rayNum = (asciiLineNum - 1) % 3
+        indexOnRay = (asciiLineNum - 1) // 3
+        if rayNum == 0:
+            # middle ray: (1,0), (0,2), (-1,4), (-2,6)
+            return 1 - indexOnRay, 2 * indexOnRay
+        elif rayNum == 1:
+            # leftmost ray: (0,1), (-1,3), (-2,5), ...
+            return -indexOnRay, 2 * indexOnRay + 1
+        else:
+            # innermost ray: (1,1), (0,3), (-1,5)
+            return 1 - indexOnRay, 2 * indexOnRay + 1
 
-    Of course, this breaks down as soon as you roll off the 120-degree
-    symmetry line towards the top of the map. Here we either have to make
-    whitespace count or insert dummy placeholders. The placeholders
-    are going to be less fragile.
+    def _makeOffsets(self):
+        """
+        One third hex grids have larger offsets at the bottom.
+        """
+        self.asciiOffsets = []
+        for li, _line in enumerate(self.asciiLines):
+            iBase, jBase = self._getIJBaseByAsciiLine(li)
+            self.asciiOffsets.append(iBase - 1)
+        self.asciiOffsets.reverse()  # since getIJ works from bottom to top
+        newOffsets = []
+
+        # renomalize the offsets to start at 0
+        minOffset = min(self.asciiOffsets)
+        for li, (line, offset) in enumerate(zip(self.asciiLines, self.asciiOffsets)):
+            newOffsets.append(offset - minOffset)
+        self.asciiOffsets = newOffsets
+
+    def _updateDimensionsFromAsciiLines(self):
+        self._ijMax = self._asciiMaxCol - 1
+        self._asciiLinesOffCorner = len(self.asciiLines[-1]) - 1
+
+    def _updateDimensionsFromData(self):
+        """Add flat-hex specific corner truncation detection"""
+        AsciiMap._updateDimensionsFromData(self)
+
+        # generally the distance from the i, 0 or 0, j index to self._ijMax
+        maxIWithData = max(i for i, j in self.asciiLabelByIndices if j == 0)
+        self._asciiLinesOffCorner = (self._ijMax - maxIWithData) * 2 - 1
+        # in jagged systems we have to also check the neighbor
+        nextMaxIWithData = max(i for i, j in self.asciiLabelByIndices if j == 1)
+        if nextMaxIWithData == maxIWithData - 1:
+            # the jagged edge is lopped off too.
+            self._asciiLinesOffCorner += 1
+
+        self._asciiMaxCol = self._ijMax + 1
+        self._asciiMaxLine = self._ijMax * 2 + 1 - self._asciiLinesOffCorner
+
+
+class AsciiMapHexFullFlatsUp(AsciiMapHexThirdFlatsUp):
     """
+    Full core flats up.
```

Comment:
Maybe something like: "Full hex ~~core~~ grid with a pair of horizontal hex edges". If we want to support non-core grids, we should try to speak more generally